### ISSUE TITLE
Async support for send receive payload

### ISSUE THREAD

**Issue by PrithviAPai** [DMTF/libspdm#2480]
Currently in SPDM Requester case libspdm triggers send callback registered when it has payload to send. 
Based on return value of the callback the receive callback registered is called subsequently. 
In many applications(example: OpenBMC etc) async way of request/response is implemented at transport layer(mctp) where request is sent and once response is received it is handled.  
The current implementation of libspdm requires response to be received before returning from send callback. 

Can we have async way of handling in SPDM Requester case ? 

Thanks,
Prithvi A Pai

**Comment by steven-bellock** [comment]
See #1049 and #2406.

**Comment by PrithviAPai** [comment]
I checked the provided links. 
I see the support only for VCA. However, other APIs are still require support is that correct ? what is the plan for supporting all the APIs ?

**Comment by jyao1** [comment]
We need some one validate and confirm that split is working, before we move too far.

@PrithviAPai , can you help to try the new API?

**Comment by PrithviAPai** [comment]
@jyao1  I tried testing this PR(https://github.com/DMTF/libspdm/pull/2406) on my setup.
The test was NOT successful. 

About my setup: 
I have libspdm tagged version 3.1.1 on top of which I cherry-picked the changes from the above PR.
SPDM version I am using isa 1.0.
I have my device which acts as SPDM Requester and i2c simulator as SPDM Responder on which spdm-emu code is running.
The transport between the two is MCTP over SMBUS.

Test Results:
1) Without this PR everything works fine.
2) We are NOT encoding/decoding transport layer messages(In case of MCTP adding SPDM_MSG_TYPE(5) to beginning of payload etc and removal of same on arrival of message)
3) The response from SPDM Emulator after removal of SPDM_MSG_TYPE(5) fails to be validated when ```libspdm_process_response_version``` is called. 
4) The receive payload seems fine  however, the check fails(attached request/response payload below).
It fails in very first check 
```
    if (spdm_response->header.spdm_version != SPDM_MESSAGE_VERSION_10) {
        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
        goto receive_done;
    }
```

Request/Response Payload:
```
spdm_responder_emu version 0.1
ver - 0x10
context_size - 0x6420
Platform server listening on port 2323
Client accepted
Platform port Receive command: 00 00 de ad
Platform port Receive transport_type: 00 00 00 01
Platform port Receive size: 00 00 00 0e
Platform port Receive buffer:
    43 6c 69 65 6e 74 20 48 65 6c 6c 6f 21 00
Platform port Transmit command: 00 00 de ad
Platform port Transmit transport_type: 00 00 00 01
Platform port Transmit size: 00 00 00 0e
Platform port Transmit buffer:
    53 65 72 76 65 72 20 48 65 6c 6c 6f 21 00
Platform port Receive command: 00 00 00 01
Platform port Receive transport_type: 00 00 00 01
Platform port Receive size: 00 00 00 05
Platform port Receive buffer:
    05 10 84 00 00
SpdmReceiveRequest[.] ...
SpdmReceiveRequest[0] (0x4):
0000: 10 84 00 00
SpdmSendResponse[0] ...
SpdmSendResponse[0] (0x8):
0000: 10 04 00 00 00 01 00 10
Platform port Transmit command: 00 00 00 01
Platform port Transmit transport_type: 00 00 00 01
Platform port Transmit size: 00 00 00 09
Platform port Transmit buffer:
    05 10 04 00 00 00 01 00 10
```

**Comment by PrithviAPai** [comment]
> We need some one validate and confirm that split is working, before we move too far.
> 
> @PrithviAPai , can you help to try the new API?

Hi @jyao1, 
I did test the APIs provided in the PR. 

About my setup: 
I have my device acting as SPDM Requester and a i2c simulator acting as SPDM Responder(which internally uses spdm-emu)
The transport used is MCTP over SMBUS.

Findings:
1) Transport layer ```encode/decode``` is NOT supported with the APIs provided. (The integrator has to add/remove Message Type SPMD(5) in case of MCTP).
2) GET VERSION Requester and Response is obtained.
3) GET CAP Request is sent but Response validation fails. The reason is:
```
    if (spdm_response->header.spdm_version != spdm_request->header.spdm_version) {
        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO, "libspdm_process_response_capabilities header request version %02x\n",spdm_request->header.spdm_version)); 
        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
        goto receive_done;
    }

Prints: 0x00
```
Last message is NOT properly assigned.

**Comment by PrithviAPai** [comment]
@steven-bellock @jyao1  can this be supported ? 
The PR added is NOT helping as transport encoding is not done and Receive part is NOT correctly handled. 
BMC is designed to work in asynchronous manner but libspdm requires synchronous approach. Please provide your feedback

**Comment by steven-bellock** [comment]
I think what you want is a library that completely removes the transport portion. So in this case the Integrator
1. Calls libspdm to generate a request that is stored in some buffer. This request may be encrypted.
2. Sends the request to the Responder.
3. Receives the response from the Responder and puts it in some buffer. This response may be encrypted.
4. Calls libspdm to parse the Response.

In particular the Integrator has full ownership over 2 and 3. If so then yes, that's doable. @JakubLedworowski does that work for you as well?

**Comment by PrithviAPai** [comment]
yes, existing library can work as it is till encode/decode transport messages. Post which we can make sure the integrator takes care of sending and receiving the API. 
This helps in allowing integrator send and receive response in async way.

**Comment by JakubLedworowski** [comment]
@steven-bellock yes, that approach works for me as long as the integrator has the option to use the transport layer methods directly (eg. libspdm_transport_mctp_encode_message) to avoid code duplication as it is already implemented in libspdm.

So, in case the integrator wants to use the new approach while reusing transport layer the flow would look like this:

1. Calls libspdm to generate a request that is stored in some buffer. This request may be encrypted.
2. [Optional] Calls libspdm to encode the request
3. Sends the request to the Responder.
4. Receives the response from the Responder and puts it in some buffer. This response may be encrypted.
5. [Optional] Calls libspdm to decode the response
6. Calls libspdm to parse the Response.

**Comment by PrithviAPai** [comment]
@steven-bellock up to transport layer encoding/decoding can be as we have currently. But, integrator need option to send/receive  without having callbacks called for the same.

**Comment by PrithviAPai** [comment]
@steven-bellock can you please tell us some rough time when this will be available so that we can plan our deliverables accordingly ? 
Thanks a lot for your support and quick reply.

**Comment by steven-bellock** [comment]
If it's at the normal pace of things maybe a year. We are focused on SPDM 1.3 features right now and it's taking a (long) while. That can obviously be accelerated if you, or someone else, implements the feature.

**Comment by PrithviAPai** [comment]
@steven-bellock please suggest if this looks.
```SEND PART```
1) Integrator to acquire ```SenderBuffer```.
2) Integrator to call ```libspdm_build_spdm_request```
3) ```libspdm_build_spdm_request``` to build spdm request + add transport encoding
4) Integrator gets the SPDM payload to be sent to spdm endpoint

```Questions```
In libspdm_handle_large_request we have a looping logic which requires callback to triggered
How should that be handled when we split the APIs ?

**Comment by raghuncstate** [comment]
I'm interested in this feature as well.

**Comment by ratagupt** [comment]
@steven-bellock : I am also interested in this feature.
